Mathlib/Data/UInt.lean

```diff
@@ -1,7 +1,178 @@
+    instance : Inhabited (Fin size) where
+      default := Fin.ofNat' 0 size_positive
+
+    lemma val_eq_of_lt : a < size → (ofNat a).val = a := Fin.val_eq_of_lt
+
+    lemma sub_def (a b : $typeName) : a - b = ⟨a.val - b.val⟩ := rfl
+
+    lemma mul_def (a b : $typeName) : a * b = ⟨a.val * b.val⟩ := rfl
+
+    lemma mod_def (a b : $typeName) : a % b = ⟨a.val % b.val⟩ := rfl
+
+    lemma add_def (a b : $typeName) : a + b = ⟨a.val + b.val⟩ := rfl
+
+    lemma eq_of_val_eq : ∀ {a b : $typeName}, a.val = b.val -> a = b
+    | ⟨f1⟩, ⟨f2⟩, h => congrArg mk h
+
+    lemma val_eq_of_eq : ∀ {a b : $typeName}, a = b -> a.val = b.val
+    | ⟨f1⟩, ⟨f2⟩, h => congrArg val h
+
+    @[simp] lemma mk_val_eq : ∀ (a : $typeName), mk a.val = a
+    | ⟨a, _⟩ => rfl
+
+    instance : AddSemigroup $typeName where
+      add_assoc := fun _ _ _ => congrArg mk (AddSemigroup.add_assoc _ _ _)
+
+    instance : AddCommSemigroup $typeName where
+      add_comm := fun _ _ => congrArg mk (AddCommSemigroup.add_comm _ _)
+
+    instance : Semigroup $typeName where
+      mul_assoc := fun _ _ _ => congrArg mk (Semigroup.mul_assoc _ _ _)
+
+    instance : CommSemigroup $typeName where
+      mul_comm := fun _ _ => congrArg mk (CommSemigroup.mul_comm _ _)
+
+    instance : AddMonoid $typeName :=
+      let add_zero_ : ∀ (a : $typeName), a + 0 = a := fun a => by
+        have h0 : a.val + (0 : $typeName).val = a.val := AddMonoid.add_zero a.val
+        simp only [add_def, h0, mk_val_eq]
+    {
+      add_zero := add_zero_
+      zero_add := fun _ => by rw [add_comm]; exact add_zero_ _
+      nsmul := fun x a => ⟨AddMonoid.nsmul x a.val⟩
+      nsmul_zero' := fun x => congrArg mk (AddMonoid.nsmul_zero' x.val)
+      nsmul_succ' := fun n x => congrArg mk (AddMonoid.nsmul_succ' n x.val)
+    }
+
+    instance : AddCommMonoid $typeName where
+      add_comm := add_comm
+
+    instance : Neg $typeName where
+      neg a := mk (-a.val)
+
+    instance : SubNegMonoid $typeName where
+      sub_eq_add_neg := fun _ _ => congrArg mk (SubNegMonoid.sub_eq_add_neg _ _)
+      gsmul := fun x a => ⟨SubNegMonoid.gsmul x a.val⟩
+      gsmul_zero' := fun a => congrArg mk (SubNegMonoid.gsmul_zero' a.val)
+      gsmul_succ' := fun m a => congrArg mk (SubNegMonoid.gsmul_succ' m a.val)
+      gsmul_neg' := fun m a => congrArg mk (SubNegMonoid.gsmul_neg' m a.val)
+
+    instance : Monoid $typeName :=
+      let mul_one_ : ∀ (a : $typeName), a * (1 : $typeName) = a := fun a => by
+        have h0 : a.val * (1 : $typeName).val = a.val := Monoid.mul_one a.val
+        simp only [mul_def, h0, mk_val_eq]
+    {
+      mul_one := mul_one_
+      one_mul := by intro a; rw [mul_comm]; exact mul_one_ a
+      npow_zero' := fun _ => rfl
+      npow_succ' := fun _ _ => rfl
+    }
+
+    instance : CommMonoid $typeName where
+      mul_comm := mul_comm
+
+    lemma val_zero : (0 : $typeName).val = (0 : Fin size) := by
+      delta OfNat.ofNat
+      delta $deltaName
+      simp only [ofNat, instOfNat, Zero.zero]
```

Comment:
```suggestion
    lemma val_zero : (0 : $typeName).val = (0 : Fin size) := rfl
```
The two zeroes here should absolutely be definitionally equal, and indeed they are.